### ISSUE TITLE
[ENH] Create a way to pre-program a workout for echelon bikes

### ISSUE THREAD

**Issue by ptrinephi** [cagnulein/qdomyos-zwift#181]
Implement a way to "program" automated resistance changes for a workout.  For example one may want 5mintues at resistance 10 then 10mn@14, 10@16, 5@10...
I'm not sure what kind of interface you want to create to enter the data but maybe for v1 read a txt file that has data like:
time@resistance<CR>
5@10
10@14
10@16
5@10

The last entry would basically mean go forever at resistance 10 or maybe when the last entry is completed set the resistance back to the users' setting.

**Comment by cagnulein** [comment]
It's already possibile in the last beta version ;)

Il lun 1 mar 2021, 20:34 3phi <notifications@github.com> ha scritto:

**Comment by ptrinephi** [comment]
Really?  I must have missed that! where are the instructions?

------ Original Message ------

**Comment by cagnulein** [comment]
https://www.facebook.com/groups/149984563348738/permalink/177286447285216/

Il lun 1 mar 2021, 21:22 3phi <notifications@github.com> ha scritto:

**Comment by ptrinephi** [comment]
Excellent, I didn't see that, thanks

------ Original Message ------

**Comment by cagnulein** [comment]
@ptrinephi did you try to use it? because i tried on the treadmill and it work well but i didn't try to the bike ;)

**Comment by ptrinephi** [comment]
I've created a simple file and uploaded it to the phone, selected the 
Open Train Program and loaded my file.  It then says "Program has been 
loaded correctly.  Press start to begin"
then nothing happens.  Resistance stays at 1.

Also, the open train program option opens any types of files and always 
says "Program has been loaded correctly.  Press start to begin"

------ Original Message ------

**Comment by cagnulein** [comment]
Did you press start and start riding? The resistance will follow your
program

Il mar 2 mar 2021, 18:56 3phi <notifications@github.com> ha scritto:

**Comment by ptrinephi** [comment]
This feature does not seem to work.  I load my program and press start 
and pedal but the resistance never changes.

------ Original Message ------

**Comment by cagnulein** [comment]
Are you using version 2.4.5? Because from the last screenshot seems not.
Anyway just send me the log if you're using the last one.
Thanks

Il gio 4 mar 2021, 06:42 3phi <notifications@github.com> ha scritto:

**Comment by ptrinephi** [comment]
Just upgraded to 2.4.5 and still not working

[debug-Wed Mar 3 21_52_44 2021.zip](https://github.com/cagnulein/qdomyos-zwift/files/6081163/debug-Wed.Mar.3.21_52_44.2021.zip)

**Comment by cagnulein** [comment]
I guess i got the issue

Warning:   QObject::connect: No such slot
echelonconnectsport::changeResistance(double) in ../src/homeform.cpp:281

I will check it today
Thanks for reporting it

Il gio 4 mar 2021, 06:56 3phi <notifications@github.com> ha scritto:

**Comment by cagnulein** [comment]
it should be fixed in version 2.4.6 @ptrinephi

**Comment by cagnulein** [comment]
@ptrinephi did you test this?

**Comment by ptrinephi** [comment]
Sorry I've been injured so no biking this entire week!

------ Original Message ------

**Comment by cagnulein** [comment]
Oh I'm sorry!
Hoping not so bad!

Il ven 12 mar 2021, 21:55 3phi ***@***.***> ha scritto:

**Comment by ptrinephi** [comment]
no worries, I go out on a real bike twice a week, last ride was a bit 
rough twisted my ankle!

------ Original Message ------

**Comment by ptrinephi** [comment]
I tried it yesterday and it worked but today it didn't.  I think I'm not 
clear on the exact process.  After selected the training program does it 
start right away?  Do I need to pause then press start to start the 
program?  Do I need to have option "pause on start" turned on?  It's not 
very clear.

Maybe it should start the program as soon as you load the training?

Also, it should complain if you load an incompatible file, right now it 
just says "press start" even when you load junk file.


------ Original Message ------
From: "Pierre" ***@***.***>
To: "cagnulein/qdomyos-zwift" 
***@***.***>; 
"cagnulein/qdomyos-zwift" ***@***.***>
Cc: "Mention" ***@***.***>
Sent: 3/12/2021 13:58:46
Subject: Re[2]: [cagnulein/qdomyos-zwift] [ENH] Create a way to 
pre-program a workout for echelon bikes (#181)

>no worries, I go out on a real bike twice a week, last ride was a bit 
>rough twisted my ankle!
>
>------ Original Message ------
>From: "Roberto Viola" ***@***.***>
>To: "cagnulein/qdomyos-zwift" ***@***.***>
>Cc: "3phi" ***@***.***>; "Mention" 
***@***.***>

**Comment by cagnulein** [comment]
Yes it will start when you press start, so I'd you already started it you
should pause and then start again.
I know that it's not very clear, this function was created only for my girl
friend ;)
I will improve it in the future
R.

Il dom 14 mar 2021, 19:56 3phi ***@***.***> ha scritto:

**Comment by ptrinephi** [comment]
It's a great feature, thanks

------ Original Message ------

**Comment by ptrinephi** [comment]
Roberto,

Allow me to make a few suggestions for this feature:

1.When opening the list of file selection, only display the files with 
extension *.xml or *.qzt for (qdomyos-zwift training).  This would make 
it easier to find the training files in the download folder full of 
debug, wip, and junk files.
2.Check the file format to make sure it's a valid QZT file. If an 
incompatible file is loaded, throw a message and prompt again.
3.On successful load, either:
     3.1.auto start the training or
     3.2.prompt the user to start the training
4.When the training is loaded, change the Zwift icon to a QZT icon to 
indicate a QZTraining is active.  (I don't think people will use zwift 
and QZTraining at the same time)
     4.1.Make the QZT icon disabled or gray when the training is not 
running
     4.2.Make theQZT icon full color when the training is active
5.When the training is complete, make the QZT icon disabled/gray.
6.When the training is active, press the QZT icon to pause the training 
and make QZT icon blink, press again to resume the training.  While QZT 
is paused the workout continues.  If the workout is paused, QZT should 
be paused as well.  Resuming the workout also resumes the training.
7.Add a global offset to the resistance.  For example, if I feel strong 
today I may want to add 2 resistance to every steps (or decrease by 2 
because my ankle is weak today!)

I know it's a lot but it's just ideas, all of which would make it a 
really nice training interface.

Thanks,

PT




------ Original Message ------

**Comment by cagnulein** [comment]
Yeah a ton of work :)
But you have definitely right. I will improve it!
For 7. You can already doing it using the plus and minus Buttons on the
Target resistance tiles


Il lun 15 mar 2021, 03:20 3phi ***@***.***> ha scritto:
